server/api/controllers/exam-bank-controller.ts

```diff
@@ -1,80 +1,170 @@
+        | UploadedFile
+        | undefined;
+
+      const {
+        department,
+        courseCode,
+        offeringTerm,
+        offeringYear,
+        type,
+      }: UploadExamBody = req.body;
+
+      const termCode = TermNameController.getTermCodeFromTermAndYear(
+        offeringTerm,
+        Number(offeringYear)
+      );
+
+      const examWithFiles: ExamWithFiles = {
+        department,
+        courseCode,
+        term: termCode.toString(),
+        type,
+        examFile: examFile ? this.examFileName(false, req.body) : undefined,
+        solutionFile: solutionFile
+          ? this.examFileName(true, req.body)
+          : undefined,
+        examFileObject: examFile,
+        solutionFileObject: solutionFile,
+      };
+
+      // add to upload queue
+      this.uploadQueue.push({
+        exam: examWithFiles,
+        errors: [],
+        res,
+      });
+    } catch (err) {
+      this.logger.error(err.message);
+      res.send(400).redirect("/admin/upload-exam");
+    }
+  }
+
+  deleteFile(req: Request, res: Response): void {
+    try {
+      const { fileName }: { fileName: string } = req.body;
+
+      if (!fileName) {
+        throw new Error("File name not provided with delete request");
+      }
+
+      const fullPath = `public/exams/${fileName}`;
+      if (!existsSync(fullPath)) {
+        throw new Error("File provdided does not exist.");
+      }
+
+      const deleteRequest: ExamDeleteRequest = {
+        fileName,
+        path: fullPath,
+        res,
+      };
+
+      this.deleteQueue.push(deleteRequest);
+    } catch (err) {
+      this.logger.error(err.message);
+      res.status(400).send(err.message);
+    }
+  }
 
-export class ExamBankController {
-  static logger = new Logger("Exam Bank Controller");
+  async processFileUpload(request: ExamUploadRequest): Promise<void> {
+    const { exam, res } = request;
 
-  static async hideExamFile(examName: string): Promise<void> {
+    try {
+      if (exam.examFile && exam.examFileObject) {
+        const examFilePath = this.getFilePath(exam.examFile);
+        await exam.examFileObject.mv(examFilePath);
+        this.logger.info(`${exam.examFile} uploaded to ${examFilePath}`);
+      }
+
+      if (exam.solutionFile && exam.solutionFileObject) {
+        const solutionFilePath = this.getFilePath(exam.solutionFile);
+        await exam.solutionFileObject.mv(solutionFilePath);
+        this.logger.info(
+          `${exam.solutionFile} uploaded to ${solutionFilePath}`
+        );
+      }
+    } catch (err) {
+      this.logger.error(err);
+    }
+
+    await this.rewriteFileJson();
+
+    res.send({ status: "success", errors: [] });
+  }
+
+  async hideExamFile(examName: string): Promise<void> {
     const currentUrl = `public/exams/${examName}.pdf`;
     const newUrl = `public/exams/${examName}-hidden.pdf`;
 
-    if (!fs.existsSync(currentUrl)) {
+    if (!existsSync(currentUrl)) {
       throw new Error("Exam not found");
```

Comment:
This is throwing an error that crashes the site when you try to hide the same exam twice in a row - want me to take a look? I think I may have caused it originally lol